### ISSUE TITLE
[Feature/Question] "Hold Ctrl to use the DefaultSandbox"

### ISSUE THREAD

**Issue by stdedos** [sandboxie-plus/Sandboxie#364]
On fresh Sandboxie installs there is this feature that will inform you "if you keep Ctrl pressed, Sandboxie will automatically select DefaultSandbox and run there, without prompt".

How is that feature implemented?:
* I have an old installation, and DefaultSandbox does not exist (or it is renamed)
* Is there a way to reverse it? (i.e. Hold Ctrl to select a sandbox, otherwise it will go to DefaultSandbox)
* Or use another key? (i.e. Shift)

Apologies for sending you code diving; hopefully it won't be dumpster diving 😀

**Comment by typpos** [comment]
> How is that feature implemented?

Application start.exe checks if CTRL is pressed, and if so, skips the "select box" dialog and runs the target application in a box called "DefaultBox". Conditions:
* CTRL pressed
* SHIFT not pressed
* No parameter to select a specific box, /box:xxx. However /box:__ask__ or /box:current are ok.
* No parameter /disable_force or /dfp
* start.exe is started from a shortcut or as a window (ie, not from a command line prompt)
* DefaultBox is not disabled

If CTRL and SHIFT are both pressed, start.exe ignores "forced programs" settings:
* https://davidxanatos.github.io/sandboxie-docs/Content/ForceProcess.html
* https://davidxanatos.github.io/sandboxie-docs/Content/ProgramStartSettings#program

Note the security risk: If you accidentally hold CTRL AND SHIFT instead of just CTRL while double-clicking a shortcut to start.exe, then the target application is executed outside the sandbox. For example, create this shortcut and try:
`"c:\Program Files\Sandboxie\Start.exe" /box:__ask__ cmd.exe`
This can be trouble if you just downloaded an untrusted executable that now sits in your download folder outside your sandbox and you try to run it in the DefaultBox from Windows File Explorer via Context-Menu > Run Sandboxed [CTRL+SHIFT]. Argh!


>  DefaultSandbox does not exist

Create a sandbox with the exact name "DefaultBox". This name is hard-coded in start.exe and a few other places. Alternatively, you can simply change it by updating the code base and creating a pull request. :-)

**Comment by stdedos** [comment]
> Note the security risk: If you accidentally hold CTRL AND SHIFT instead of just CTRL while double-clicking a shortcut to start.exe, then the target application is executed outside the sandbox. For example, create this shortcut and try:
> `"c:\Program Files\Sandboxie\Start.exe" /box:__ask__ cmd.exe`
> This can be trouble if you just downloaded an untrusted executable that now sits in your download folder outside your sandbox and you try to run it in the DefaultBox from Windows File Explorer via Context-Menu > Run Sandboxed [CTRL+SHIFT]. Argh!

So, that sounds like a bug to me?
Or Sandboxie has had specifically mapped:
```bash
case $keys in
	Ctrl+Shift) DFP;;
	Ctrl) DefaultBox;;
    *) :;;
esac
```
?

If so, should we file an issue for that?

**Comment by typpos** [comment]
> So, that sounds like a bug to me?

It's intended behaviour, so not a coding bug:
https://github.com/sandboxie-plus/Sandboxie/blob/e5534f4ed1ce14f7f99c469014b1c51130ebbb9c/Sandboxie/apps/start/start.cpp#L1328-L1339

However in my view it's a clear usability and security-focus design bug. The idea of using SHIFT as a convenience feature overrides the much more important objective of security. When you select "Run sandboxed" from the context menu, there should be no way to silently undo that by pressing SHIFT to have it not sandboxed.

Similarly, the idea of "CTRL for default box" would better be implemented by offering a second context menu "Run in default sandbox" instead of a magic keyboard key. This would also make obsolete the strange code that ensures this behaviour is only triggered when start.exe is not run from a command prompt.

Besides, SHIFT and CTRL are unreliable in admin work with remoting and VM envs. . I've had regular instances where my remote/virtual sessions had the keyboard state for control keys toggled. Of course, admins NEVER run suspicious software...

> If so, should we file an issue for that?

Go for it, or rename the issue title. Up to you.

**Comment by stdedos** [comment]
I disagree a little bit: Offering more right-click entries is not really the way to go, because of the cluttering (ofc, they could be nested).

I'd solve this differently:
```cpp
KeyState = GetAsyncKeyState(VK_CONTROL); 
if (KeyState & 0x8000) { 
    if (si.dwFlags & (STARTF_USESHOWWINDOW | STARTF_TITLEISLINKNAME)) { 
            auto_select_default_box = TRUE; 
    } 
} else {
    KeyState = GetAsyncKeyState(VK_SHIFT); 
    if (KeyState & 0x8000) 
        disable_force_on_this_program = TRUE; 
}
```

You can force non-sandboxing by *only* pressing Shift (if someone would like to use it), and you can still press `Ctrl` to start DefaultBox. It doesn't make sense to `disable_force_on_this_program` if you want to `auto_select_default_box`. At worst, `auto_select_default_box` when you really meant `disable_force_on_this_program` should be, at maximum, re-running the program. You should be able to detect you ran a program sandboxed quite easily (Yellow border, `[#]`)

**Comment by typpos** [comment]
Nested menu options would be acceptable to me. Bonus points for customizable context menus like in 7-zip.

SHIFT vs SHIFT+CTRL: Either. I never use them. Just don't get the existing user base confused.

Your code is better and I agree with your reasoning.

I dislike these SHIFT/CTRL options in principle. They're traps.  Security first; convenience second. As a minimum, allow users to disable this behaviour in global settings.

**Comment by stdedos** [comment]
> I dislike these SHIFT/CTRL options in principle. They're traps. Security first; convenience second. As a minimum, allow users to disable this behaviour in global settings.

I agree too. I have tried to use Ctrl to skip the dialog, but I could miss the timing.
But you may need a convenient way to DFP (I don't know of any) if you are not a malware researcher (you could argue "just move to another directory duuude" though).

However, with my solution, it's two birds with one stone: You cannot "just mispress" Shift. If you mispress Ctrl+Shift, I am protecting you right now (that's why kinda I would want to keep the behavior).

A Global disable/enable flag would be nice for sure!

**Comment by DavidXanatos** [comment]
So whats the verdict what should I do with that options?

**Comment by stdedos** [comment]
From minimum work to most work my impression / improvement suggestions on this are:

1. Don't do anything, I'll get around doing a PR this month
1. Add some note regarding CTRL + DefaultBox on the manual you maintain
1. Convert this https://github.com/sandboxie-plus/Sandboxie/blob/e5534f4ed1ce14f7f99c469014b1c51130ebbb9c/Sandboxie/apps/start/start.cpp#L1328-L1339 to that
	  > ```c++
	  > KeyState = GetAsyncKeyState(VK_CONTROL); 
	  > if (KeyState & 0x8000) { 
	  >     if (si.dwFlags & (STARTF_USESHOWWINDOW | STARTF_TITLEISLINKNAME)) { 
	  >             auto_select_default_box = TRUE; 
	  >     } 
	  > } else {
	  >     KeyState = GetAsyncKeyState(VK_SHIFT); 
	  >     if (KeyState & 0x8000) 
	  >         disable_force_on_this_program = TRUE; 
	  > }
	  > ```
1. A Global disable/enable flag [for `disable_force_on_this_program`] would be nice for sure!

And for some "food for thought" ideas:
1. Make CTRL behavior reversible (i.e. "Press Ctrl to select a sandbox otherwise run on the default one); partially conflicts with https://github.com/sandboxie-plus/Sandboxie/issues/364#issuecomment-758912646
1. Make DefaultBox a macro
1. Make DefaultBox configurable (as Sanboxie.ini setting and everything)

**Comment by typpos** [comment]
Mostly fine with stdedos summary, except regards code change:

1. Move the dwFlags test to the top so SHIFT (just like CONTROL) is also not triggered in such a call context.
2. Don't treat SHIFT+CTRL same as CTRL. SHIFT+CTRL should do nothing.

**Comment by stdedos** [comment]
> 1. Move the dwFlags test ...

Is there layman's documentation on "what are these"?
If you guys think "that's okay", then sure

> 2. Don't treat SHIFT+CTRL same as CTRL. SHIFT+CTRL should do nothing.

My code swap suggestion will ensure that. In any case, it would be better if you could add a "unit test" to ensure that "abstract code path" stays like this after fixing.

**Comment by typpos** [comment]
Now you made me code!  :-)

Like so:

```c++
if (si.dwFlags & (STARTF_USESHOWWINDOW | STARTF_TITLEISLINKNAME)) {
    BOOL Alt =   GetAsyncKeyState(VK_MENU) < 0;
    BOOL Ctrl =  GetAsyncKeyState(VK_CONTROL) < 0;
    BOOL Shift = GetAsyncKeyState(VK_SHIFT) < 0;
    if(Ctrl && !Alt && !Shift)
        auto_select_default_box = TRUE;
    else if(Shift && !Alt && !Ctrl)
        disable_force_on_this_program = TRUE;
}
```

**Comment by DavidXanatos** [comment]
> 
> 
> From minimum work to most work my impression / improvement suggestions on this are:
> 
>     1. Don't do anything, I'll get around doing a PR this month
> 

how is the pull request going?

**Comment by stdedos** [comment]
> > 
> > 
> > From minimum work to most work my impression / improvement suggestions on this are:
> > 
> >     1. Don't do anything, I'll get around doing a PR this month
> > 
> 
> how is the pull request going?

Too much coding in other places 😕 

I guess from my place, it will have to spill for next month

**Comment by DavidXanatos** [comment]
In the spirit of boosting engagement I'll wait for a Pull Request on this issue however logn it may take ;)
